pytorch_lightning/trainer/trainer.py

```diff
@@ -1141,7 +1141,7 @@ def call_teardown_hook(self, model: LightningModule) -> None:
         state = self._teardown_state
 
         if self.datamodule is not None:
-            called = getattr(self.datamodule, f'has_teardown_{state}')
+            called = hasattr(self.datamodule, f'has_teardown_{state}') and getattr(self.datamodule, f'has_teardown_{state}')
```

Comment:
```suggestion
            called = getattr(self.datamodule, f'has_teardown_{state}', False)
```

Reply:
@carmocca what's the expected behavior if `state` is None? to call teardown with stage=None? 

i'm not sure why the state for teardown can be optionally None but for setup its guaranteed as non-None

Reply:
> what's the expected behavior if state is None? to call teardown with stage=None?

yes, meaning run teardown for all stages

> i'm not sure why the state for teardown can be optionally None but for setup its guaranteed as non-None

It's _never_ `None` for `setup` because there's almost no code between where we set the `running_stage` and where we call setup.

But for teardown, there are a few pieces that modify it.

The best fix would be which piece incorrectly sets it to `None` and fix that instead.